### PR TITLE
Feature/change patch loc

### PR DESCRIPTION
**Pull request by snudler6**
solves #24

### CODE REVIEW

File: src/time_travel/patchers/base_patcher.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    
+    def get_patched_module(self):
+        """Do more stuff."""
+        raise NotImplementedError()
+        
+    def get_patch_actions(self):
+        """Do stuff."""
+        raise NotImplementedError()
+        
+    def start_extra_actions(self):
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: MAGIL

File: docs/index.md [hunk 2]
```diff
@@ -122,3 +122,14 @@ setup(
     }
 )
 ```
+
+The class should do 1 of 2 things:
+
+1. Implement it's own `start` and `stop` of the patcher
```
Comment (FarmerSez) [inline docs/index.md]: Implement -> Override

File: src/time_travel/patchers/base_patcher.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+"""Base class for patching time modules."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: time and I/O events?

File: src/time_travel/patchers/base_patcher.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+class BasePatcher(object):
+    """Base class for patching time modules."""
+    
+    def __init__(self,
+                 clock,
+                 event_pool=None,
+                 modules_to_patch=None,
+                 patcher_module=None):
+        """Create the patch."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: Create the patch**er**.

File: src/time_travel/patchers/base_patcher.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+import mock
+import inspect
+
+
+class BasePatcher(object):
+    """Base class for patching time modules."""
+    
+    def __init__(self,
+                 clock,
+                 event_pool=None,
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: I don't think this should ever be None. If it is - it means that it's not initialized correctly by the main class. Remove the default argument pl0x.

File: src/time_travel/patchers/base_patcher.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        """Return the namespace of the patcher's events."""
+        return None
+    
+    @classmethod
+    def get_events_types(cls):
+        """Return Enum of the patcher's events types."""
+        return None
+    
+    def get_patched_module(self):
+        """Do more stuff."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: Docstring.

File: src/time_travel/patchers/base_patcher.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        raise NotImplementedError()
+        
+    def get_patch_actions(self):
+        """Return list of patches to do.
+        
+        The list structure is tuples containgin:
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: containing (typo)

File: src/time_travel/patchers/base_patcher.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            # not given a specific module to patch on.
+            # Create the list of all modules to search for the patched objects.
+            # Patch on all loaded modules.
+            modules = [
+                module for mod_name, module in sys.modules.items() if
+                inspect.ismodule(module) and
+                hasattr(module, '__name__') and
+                # Don't patch inside this module,
+                # or inside the original module.
+                module.__name__ not in ([patched_module, self.patcher_module,
```
Comment (FarmerSez) [inline src/time_travel/patchers/base_patcher.py]: move `self.patcher_module` to a new line as well.

File: src/time_travel/patchers/datetime_patcher.py [hunk 9]
```diff
@@ -167,88 +143,36 @@ class DatetimePatcher(BasicPatcher):
...
-        super(DatetimePatcher, self).__init__(**kwargs)
+        super(DatetimePatcher, self).__init__(patcher_module=__name__,
+                                              **kwargs)
         
         FakeDate._now = self._now
         FakeDatetime._now = self._now
         
-        self._undo_set = set()
+    def get_patched_module(self):
+        """Do more stuff."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: Docstring

File: src/time_travel/patchers/datetime_patcher.py [hunk 10]
```diff
@@ -167,88 +143,36 @@ class DatetimePatcher(BasicPatcher):
...
+        return datetime
         
-    def start(self):
-        """Start the patch of datetime.datetime class.
+    def get_patch_actions(self):
+        """Do stuff."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/datetime_patcher.py]: Docstring

File: src/time_travel/patchers/poll_patcher.py [hunk 11]
```diff
@@ -1,9 +1,10 @@
 """A patch to the select.poll object."""
 
-from .basic_patcher import BasicPatcher
+from .base_patcher import BasePatcher
 
 import select as select_lib
 import mock
+import select
```
Comment (FarmerSez) [inline src/time_travel/patchers/poll_patcher.py]: Remove.

File: src/time_travel/patchers/poll_patcher.py [hunk 12]
```diff
@@ -86,14 +87,20 @@ def _is_relevant_fd_event(fd, evt):
...
     """Patcher for select.poll."""
     
     def __init__(self, *args, **kwargs):
         """Create the patch."""
         super(PollPatcher, self).__init__(*args, **kwargs)
 
-        self.patches = [mock.patch('select.poll', self._mock_poll)]
+    def get_patched_module(self):
+        """Do more stuff."""
+        return select
```
Comment (FarmerSez) [inline src/time_travel/patchers/poll_patcher.py]: `return select_lib`

File: src/time_travel/patchers/select_patcher.py [hunk 13]
```diff
@@ -1,13 +1,14 @@
 """A patch to the select.select function."""
 
-from .basic_patcher import BasicPatcher
+from .base_patcher import BasePatcher
 
 import select as select_lib
 from enum import Enum
 import mock
+import select
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: Use `select_lib` instead.

File: src/time_travel/patchers/poll_patcher.py [hunk 14]
```diff
@@ -86,14 +87,20 @@ def _is_relevant_fd_event(fd, evt):
...
+class PollPatcher(BasePatcher):
     """Patcher for select.poll."""
     
     def __init__(self, *args, **kwargs):
         """Create the patch."""
         super(PollPatcher, self).__init__(*args, **kwargs)
 
-        self.patches = [mock.patch('select.poll', self._mock_poll)]
+    def get_patched_module(self):
+        """Do more stuff."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/poll_patcher.py]: Docstring

File: src/time_travel/patchers/select_patcher.py [hunk 15]
```diff
@@ -21,7 +22,13 @@ def __init__(self, *args, **kwargs):
             side_effect=self._mocked_select,
             spec=select_lib.select)
         
-        self.patches = [mock.patch('select.select', self.select)]
+    def get_patch_actions(self):
+        """Return generator containing all patches to do."""
+        return [('select', select.select, self.select)]
+        
+    def get_patched_module(self):
+        """Do stuff."""
```
Comment (FarmerSez) [inline src/time_travel/patchers/select_patcher.py]: Docstring

File: src/time_travel/patchers/time_patcher.py [hunk 16]
```diff
@@ -1,27 +1,34 @@
...
+    def get_patched_module(self):
+        """Do more stuff."""
+        return time
         
-        self.patches.append(
-            mock.patch('time.sleep', side_effect=self._advance_time_stamp))
+    def get_patch_actions(self):
+        """Return generator containing all patches to do."""
+        return [
+            ('time', time.time, mock.Mock(side_effect=self._get_timestamp)),
```
Comment (FarmerSez) [inline src/time_travel/patchers/time_patcher.py]: You don't need to use Mock here at all. You need to override the original functions with the `side effects` functions themselves.

File: docs/index.md (hunk 2 above)
Comment (snudler6) [reply docs/index.md]: boop

File: src/time_travel/patchers/base_patcher.py (hunk 5 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: Done.

File: src/time_travel/patchers/base_patcher.py (hunk 4 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: gloop.

File: src/time_travel/patchers/base_patcher.py (hunk 6 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: I'm with stupid.

File: src/time_travel/patchers/base_patcher.py (hunk 7 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: done.

File: src/time_travel/patchers/base_patcher.py (hunk 8 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: Don't have to according to pep8, but OK.

File: src/time_travel/patchers/datetime_patcher.py (hunk 9 above)
Comment (snudler6) [reply src/time_travel/patchers/datetime_patcher.py]: Ditto.

File: src/time_travel/patchers/datetime_patcher.py (hunk 10 above)
Comment (snudler6) [reply src/time_travel/patchers/datetime_patcher.py]: Ditto (Female).

File: src/time_travel/patchers/poll_patcher.py (hunk 11 above)
Comment (snudler6) [reply src/time_travel/patchers/poll_patcher.py]: Heard you talk about it.

File: src/time_travel/patchers/poll_patcher.py (hunk 14 above)
Comment (snudler6) [reply src/time_travel/patchers/poll_patcher.py]: Ditto (Both)

File: src/time_travel/patchers/poll_patcher.py (hunk 12 above)
Comment (snudler6) [reply src/time_travel/patchers/poll_patcher.py]: Done before I even read this.

File: src/time_travel/patchers/select_patcher.py (hunk 13 above)
Comment (snudler6) [reply src/time_travel/patchers/select_patcher.py]: So predictable.

File: src/time_travel/patchers/select_patcher.py (hunk 15 above)
Comment (snudler6) [reply src/time_travel/patchers/select_patcher.py]: Ditto

File: src/time_travel/patchers/time_patcher.py (hunk 16 above)
Comment (snudler6) [reply src/time_travel/patchers/time_patcher.py]: Removed!

File: src/tests/test_poll.py [hunk 17]
```diff
@@ -22,7 +22,8 @@ def setup_method(self, method):
         """Start a poll patcher"""
         self.event_pool = EventPool()
         self.clock = TimeMachineClock(clock_listeners=[self.event_pool])
-        self.patcher = PollPatcher(self.clock, self.event_pool)
+        self.patcher = PollPatcher(self.clock, self.event_pool,
```
Comment (FarmerSez) [inline src/tests/test_poll.py]: event pool in new line

File: src/tests/test_select.py [hunk 18]
```diff
@@ -16,7 +16,8 @@ def setup_method(self, method):
 
         self.clock = TimeMachineClock(clock_listeners=[self.event_pool])
         
-        self.patcher = SelectPatcher(self.clock, self.event_pool)
+        self.patcher = SelectPatcher(self.clock, self.event_pool,
```
Comment (FarmerSez) [inline src/tests/test_select.py]: same

File: src/time_travel/patchers/base_patcher.py (hunk 3 above)
Comment (snudler6) [reply src/time_travel/patchers/base_patcher.py]: Done.

**Pull request merged**